libp2p/NodeTable.cpp

```diff
@@ -262,57 +263,29 @@ void NodeTable::doDiscoveryRound(
         });
 }
 
-vector<shared_ptr<NodeEntry>> NodeTable::nearestNodeEntries(NodeID _target)
+vector<shared_ptr<NodeEntry>> NodeTable::nearestNodeEntries(NodeID const& _target)
 {
-    // send s_alpha FindNode packets to nodes we know, closest to target
-    static unsigned lastBin = s_bins - 1;
-    unsigned head = distance(m_hostNodeID, _target);
-    unsigned tail = head == 0 ? lastBin : (head - 1) % s_bins;
-
-    map<unsigned, list<shared_ptr<NodeEntry>>> found;
+    auto const distanceToTargetLess = [](pair<int, shared_ptr<NodeEntry>> const& _node1,
+                                          pair<int, shared_ptr<NodeEntry>> const& _node2) {
+        return _node1.first < _node2.first;
+    };
+
+    std::multiset<pair<int, shared_ptr<NodeEntry>>, decltype(distanceToTargetLess)>
+        nodesByDistanceToTarget(distanceToTargetLess);
+    for (auto const& bucket : m_buckets)
+        for (auto const& nodeWeakPtr : bucket.nodes)
+            if (auto node = nodeWeakPtr.lock())
+            {
+                nodesByDistanceToTarget.emplace(distance(_target, node->id()), node);
```

Comment:
This is currently very inefficient, because it calculates `sha3(m_hostNodeID)` many times in the loop and `sha3(node->id())` every time `nearestNodeEntries` is called.

I will optimize it in the next PR by calculating hashes only once and making distance function get hashes as input.

Reply:
https://github.com/ethereum/aleth/pull/5586